### ISSUE TITLE
Release hook as counterpart of init hook

### ISSUE THREAD

**Issue by drobiazko** [elasticio/sailor-nodejs#62]
It would be nice to have a `release` hook which is invoked:

* after Sailor disconnected from queues  
* before the step is removed from Mesos/k8s

**Comment by zubairov** [comment]
What is the use-case for that?

**Comment by drobiazko** [comment]
Our dutch customer asked for that. They want to send an HTTP request once a step finished executing and is about to be shut down.

It would also help to implement aggregation of multiple messages temporarily stored in a bucket. For example our CSV write could close the stream inside `release` as we are sure now that we received all the messages.

**Comment by zubairov** [comment]
Good point about the aggregation, however when queues are gone you won't be able to do much, right?

**Comment by drobiazko** [comment]
Not if you put the messages into an external bucket.

**Comment by zubairov** [comment]
hm.. do you think it would make sense to call ``release`` when all incoming messages are consumed and acked but before outgoing queue is not yet closed. Otherwise we lose many possibilities e.g. snapshots, error reporting ,etc

**Comment by drobiazko** [comment]
I think we could to it that way:

1. unbind  from the data queue in order to make sure that no messages are consumed anymore
2. invoke `release`. In case of errors, report them to the errors queue
3. disconnect completely

**Comment by zubairov** [comment]
+1 can we also make sure that ``release`` is called form the context of ``this`` so that I could:

- [ ] Emit ``data``
- [ ] Emit ``error``
- [ ] Emit ``snapshot``

Would be also good to agree on - if ``release`` returns a ``Promise`` then:

- [ ] If promise will reject, ``sailor`` should emit ``error`` event
- [ ] If promise will fulfill, and result is not ``undefined`` then ``sailor`` should emit ``data`` event with result as message body
